### ISSUE TITLE
Merge Nemoh's commit 115 in Capytaine

### ISSUE THREAD

**Issue by izabala123** [mancellin/capytaine#20]
Dear Matthieu

I have had the opportunity of comparing Capytaine with Wamit and Nemoh version 115, for the DeepCWind wind turbine floating semisub platform ([https://community.ieawind.org/HigherLogic/System/DownloadDocumentFile.ashx?DocumentFileKey=12510238-c40b-9597-a845-580276c1b88d](https://community.ieawind.org/HigherLogic/System/DownloadDocumentFile.ashx?DocumentFileKey=12510238-c40b-9597-a845-580276c1b88d)) and it seems that the difference in results worth to include Nemoh irregular frequencies removal technology into Capytaine.
![View of the platform](https://github.com/izabala123/BEMRosetta/blob/6498c76bc694d4809938f0eab380c7ce0482684f/other/md%20resources/Other/tempdcw.png)
![Added mass in heave comparison from BEMRosetta](https://github.com/izabala123/BEMRosetta/blob/6498c76bc694d4809938f0eab380c7ce0482684f/other/md%20resources/Other/tempheave.jpg).
The legend means:
- "Capyta": Capytaine
- "Joined files": Nemoh v115
- "dcw10": Wamit

I can submit all the files necessary to reproduce this case test.

All the best
Iñaki Zabala

**Comment by izabala123** [comment]
Of course yes. However I will need some days to prepare the binary and run the simulation.
In parallel, could you give me the files lacked in revision 118?

**Comment by izabala123** [comment]
Dear Matthieu

These are the results of Nemoh  v113, Capytaine and Nemoh v115. 
It would be great if Capytaine could include v115 improvements. 

![Radiation damping 1 heave 3](https://user-images.githubusercontent.com/38589221/66680152-230c1f00-ec70-11e9-8de0-872238bab336.jpg)
![Added mass 1 heave 3](https://user-images.githubusercontent.com/38589221/66680169-2b645a00-ec70-11e9-9de7-a0cf8d8e69e4.jpg)

All the best
Iñaki Zabala

**Comment by mancellin** [comment]
Thank you very much. There seems to be a significant discrepancy between Nemoh and Capytaine on this test case. I'll need to investigate that...

I have uploaded the version you've asked for: https://ancell.in/tmp/2019-10-11-Nemoh_UCD_89518c8a6edc10e601437fbe88b11da3.zip (I haven't run it for years, so I can't guarantee it will work fine...).

**Comment by izabala123** [comment]
Thank you Matthieu. I have compiled and run your version 118, but as with the Mercurial one, it crashes in SOLVE_POTENTIAL_DIRECT/LISC/EXPORS, when trying to do a CLOG() with 0 in INITIALIZE_GREEN_2.f90, line 231. As you were one of the original developers, I leave to you to decide if it is worth trying to solve the problem in this version. For now version 115 seem to be the last valid one.

**Comment by izabala123** [comment]
I have used a new mesh in Capytaine with 10937 panels instead of the original 7089, and the results are almost the same:
![image](https://user-images.githubusercontent.com/38589221/66718246-f4ae5100-ede1-11e9-9f5a-c15723c719d1.png)
7089 panels
![image](https://user-images.githubusercontent.com/38589221/66718273-4951cc00-ede2-11e9-9cfd-6e310341ed30.png)
10937 panels
![image](https://user-images.githubusercontent.com/38589221/66718278-579fe800-ede2-11e9-8b3f-0564bf5a4a04.png)

**Comment by mancellin** [comment]
No, I don't think it is worthwhile to fix the commit 118, it is mostly an early draft of Capytaine.

The priority should be to figure out why Nemoh 113 and Capytaine give different results. I'll investigate that in a few days when I have some time. Thank you Iñaki for the mesh convergence test, it will be useful.

**Comment by izabala123** [comment]
I agree. I can give you all the support you require to underline the differences between 113 and 115. I am at your disposal.

**Comment by mancellin** [comment]
If you want, you can do more comparisons (capytaine vs. 113 vs. 115) on other simpler geometries. It would help to figure out if there is a particular feature in the shape of the body that causes the discrepancy. In particular, there should not be any irregular frequency for a fully immersed body.

**Comment by izabala123** [comment]
Perfect. Could you supply the geometries?

**Comment by mancellin** [comment]
Only in a couple of days, when I'm back in office.

**Comment by mancellin** [comment]
I just noticed your test case is in finite depth. Would you mind running the same comparison in infinite depth?

**Comment by mancellin** [comment]
Never mind, I was able to reproduce your results and the behavior seems to be the same in finite and infinite depth, with Capytaine 1.0 and 1.1, for direct and iterative linear solver.

**Comment by izabala123** [comment]
Do you still want the results? Capytaine simulation is over a lot of hours ago but Nemoh is still running with 8 cores ;-)

**Comment by mancellin** [comment]
You can let it finish. We'll have a reference result from Nemoh for this case. I only tested a few frequencies in Capytaine.

**Comment by izabala123** [comment]
![image](https://user-images.githubusercontent.com/38589221/67091178-483ae900-f1ac-11e9-8922-c3608fc61d57.png)
![image](https://user-images.githubusercontent.com/38589221/67091192-538e1480-f1ac-11e9-803a-edc15a06bfd8.png)
![image](https://user-images.githubusercontent.com/38589221/67091201-5852c880-f1ac-11e9-995e-8987e0bc806d.png)
![image](https://user-images.githubusercontent.com/38589221/67091210-5d177c80-f1ac-11e9-97ed-097888aae467.png)
![image](https://user-images.githubusercontent.com/38589221/67091219-63a5f400-f1ac-11e9-90ed-4ef30845ba40.png)
![image](https://user-images.githubusercontent.com/38589221/67091226-67d21180-f1ac-11e9-8cf5-948223fbc4b8.png)
![image](https://user-images.githubusercontent.com/38589221/67091232-6b659880-f1ac-11e9-89fe-b8d1c98b686f.png)
![image](https://user-images.githubusercontent.com/38589221/67091244-6f91b600-f1ac-11e9-817d-2b3864150da5.png)
![image](https://user-images.githubusercontent.com/38589221/67091322-a2d44500-f1ac-11e9-9215-e080be774d24.png)

**Comment by mancellin** [comment]
Thank you Iñaki for these results. I guess the version of Nemoh you used is commit 115?
Since the behavior seems to be the same in finite and infinite depth, let's focus on the infinite depth case which is slightly faster to compute and easier to debug.

I tried to compute the free surface for a few frequencies.
Below the radiated wave for a heaving body at omega = 2.1 rad/s and 3.0 rad/s (where major discrepancies are observed) :
![2 10](https://user-images.githubusercontent.com/31126826/67189467-92131180-f3ee-11e9-9eae-ac1e9ae1305d.gif)![3 00](https://user-images.githubusercontent.com/31126826/67190971-2aaa9100-f3f1-11e9-8436-be09ba1a170e.gif)
The resonances inside the cylinders might be a symptom of irregular frequencies. However, I'm not very familiar with this question and it still puzzles me that it would 1) affect such a wide range of frequencies and 2) give different biases between Capytaine and Nemoh 113.

From what I understood from commit 115, it is mostly an increase of resolution in the evaluation of the Green function. I need to figure out if there is actually something about irregular frequencies in there.

**Comment by izabala123** [comment]
Curious videos. For comparison I have always used v115 as a reference. To clarify it even more, I created [this Nemoh fork](https://github.com/izabala123/Nemoh) including from all from binaries to sources ready to be compiled and the demo.
If you want I can prepare a simple report with the main differences between v113 and v115.

**Comment by mancellin** [comment]
So, I dived into the mess of Nemoh branches. Actually, the different experimental branches (such as 103 and 115) share a lot of modifications in common. They might just be different variants of the same work by G. Delhommeau.

I found the following differences between the main version of Nemoh (commit 113) and commit 115:

1) More points used for the computation of the tabulated integrals (5001 versus 251)
    -> Already modifiable by the user in Capytaine. From my experience, has very little effect.

2) More points in the table of the tabulated integrals (124×676 instead of 46×328)
   -> I had in my long-term plans to let the user chose the fineness of the tabulation, but since most of the tabulation access is hard-coded, it is a bit of work to implement it cleanly. A less ambitious implementation of only the 124×676 tabulation besides the 46×328 could be easy.

3) Interpolation in the table with a fifth order polynomial instead of third order.
   -> Easy to implement in Capytaine.

4) 4-points Gauss quadrature for the integration of the Green function on a face, instead of 1-point approximation.
    -> Relatively easy to implement in Capytaine. Could be fun to compare all the quadrature implemented in [Quadpy](https://github.com/nschloe/quadpy)

5) Irregular frequencies removal.
    Since it requires the user to mesh the intersection of the body with the free surface (see [Babarit and Delhommeau 2015]), I'm not sure if it was actually active for the results presented previously in this thread.
    Unlike the previous points, it is a qualitative change in what is computed, not just a more accurate version of the same. Not sure how hard it would be to implement.

Implementing points 2 to 4 is a nice work program, but I don't know when I'll have the time to do it. However, it is easy enough to be done by new contributors (with some guidance from me). So if anybody wants to volunteer, please let me know.

Regarding point 5: @izabala123, can I let you check if your Nemoh 115 gives different results when the intersection of the free surface with the body is meshed (see e.g. half-sphere case in [Babarit and Delhommeau 2015])?

Note that none of this explains the discrepancy between Nemoh 113 and Capytaine, which is what worries me the most.

**Comment by izabala123** [comment]
Good review.  I have a rough idea of what you're asking of me. However, please I would be grateful if you could detail more what you mean with "the intersection of the free surface with the body is meshed". In addition, does [Babarit and Delhommeau 2015] is [this](https://www.researchgate.net/publication/282288245_Theoretical_and_numerical_aspects_of_the_open_source_BEM_solver_NEMOH)?

**Comment by mancellin** [comment]
The method used for the removal of irregular frequencies requires to 
mesh the z=0 plane inside the body, where the weird resonances inside 
the cylinder are occurring on the animation above.
I am referring to the proceeding paper related to the EWTEC presentation 
you linked.

**Comment by izabala123** [comment]
I have got it. In a day the results will be ready.
![image](https://user-images.githubusercontent.com/38589221/67401676-c84bbf00-f5af-11e9-81f3-895277f65519.png)

**Comment by izabala123** [comment]
Nemoh115 is running but Capytaine says multiple times:
WARNING:        The mesh of the body .\mesh\DCW10_7000Closed.dat is not inside the domain.
                                Check the values of free_surface and sea_bottom
                                or use body.keep_immersed_part() to clip the mesh.

And finally it stops with this message:
INFO:   Solve DiffractionProblem(body=.\mesh\DCW10_7000Closed.dat, omega=0.500, depth=inf, wave_direction=0.000, g=9.8066, rho=1025.0).
 Error: Impossible to compute the wave part of the Green function (division by zero).
ERROR STOP
(base) PS C:\Capyta DCW_7000 Inf Closed>

Z values of points in surface are infinitesimal, like, 9.658124e-17

**Comment by mancellin** [comment]
You can try, but I can't promise the result will make any sense.
Unlike Nemoh 115, faces at the free surface were not included in the 
logic of Capytaine.

**Comment by izabala123** [comment]
![image](https://user-images.githubusercontent.com/38589221/67548736-df4ef600-f702-11e9-8e21-518f2f820b91.png)
![image](https://user-images.githubusercontent.com/38589221/67548794-0efdfe00-f703-11e9-8139-cb3c25439fb0.png)
![image](https://user-images.githubusercontent.com/38589221/67548802-18876600-f703-11e9-85bd-2a7618c87fba.png)
![image](https://user-images.githubusercontent.com/38589221/67548823-263ceb80-f703-11e9-9256-0ea3d82613bf.png)
![image](https://user-images.githubusercontent.com/38589221/67548862-453b7d80-f703-11e9-9060-e98aee81559d.png)
![image](https://user-images.githubusercontent.com/38589221/67548867-48cf0480-f703-11e9-8cdc-30b6bc37d79d.png)
![image](https://user-images.githubusercontent.com/38589221/67548877-4cfb2200-f703-11e9-8f16-9f19e97a96a1.png)
![image](https://user-images.githubusercontent.com/38589221/67548883-51273f80-f703-11e9-82ba-e82d093214b4.png)
![image](https://user-images.githubusercontent.com/38589221/67549015-921f5400-f703-11e9-96fb-b8ecd00840fe.png)
![image](https://user-images.githubusercontent.com/38589221/67549035-9ba8bc00-f703-11e9-9764-1bfde96ac5b1.png)

**Comment by izabala123** [comment]
Good comment. Here they are with WAMIT non infinite depth. Legends mean:
- Nemoh v115 infinite depth closed mesh in free surface: DCW_7000_v115 Inf Closed
- Nemoh v115 infinite depth: DCW_7000 Inf
- Wamit non infinite depth: dcw10
![image](https://user-images.githubusercontent.com/38589221/67557108-b6cff780-f714-11e9-84a4-373595055d3d.png)
![image](https://user-images.githubusercontent.com/38589221/67557118-bafc1500-f714-11e9-9ac0-ed80ea453de6.png)
![image](https://user-images.githubusercontent.com/38589221/67557125-bf283280-f714-11e9-9a69-578a85351a77.png)
![image](https://user-images.githubusercontent.com/38589221/67557134-c4857d00-f714-11e9-8431-224e461fa744.png)
![image](https://user-images.githubusercontent.com/38589221/67557161-d1a26c00-f714-11e9-8a7b-44c981555e74.png)
![image](https://user-images.githubusercontent.com/38589221/67557167-d535f300-f714-11e9-85a2-89fe1f48afe2.png)
![image](https://user-images.githubusercontent.com/38589221/67557188-e3840f00-f714-11e9-8e7b-867e8e927c89.png)
![image](https://user-images.githubusercontent.com/38589221/67557194-e7179600-f714-11e9-884e-f362e1a8f3c9.png)
![image](https://user-images.githubusercontent.com/38589221/67557221-f4348500-f714-11e9-8a34-ee1b0703bce6.png)
![image](https://user-images.githubusercontent.com/38589221/67557234-fbf42980-f714-11e9-980b-4ac1f987ef2c.png)

**Comment by izabala123** [comment]
In fact this may contradict the assertion that closing the mesh reduces irregular frequencies, as unclosed results are more similar to Wamit than closed ones.
However, for a more reliable comparison, I have asked that the same simulation be launched with Wamit and infinite depth.

**Comment by ffrancesco** [comment]
Hi @mancellin and @izabala123, is there any followup on this thread? I´m using Capytaine and i might be able to help out to implement point 2 and point 4 as quoted below if @mancellin is up to provide some directions. 


> So, I dived into the mess of Nemoh branches. Actually, the different experimental branches (such as 103 and 115) share a lot of modifications in common. They might just be different variants of the same work by G. Delhommeau.
> 
> I found the following differences between the main version of Nemoh (commit 113) and commit 115:
> 
> 1. More points used for the computation of the tabulated integrals (5001 versus 251)
>    -> Already modifiable by the user in Capytaine. From my experience, has very little effect.
> 2. More points in the table of the tabulated integrals (124×676 instead of 46×328)
>    -> I had in my long-term plans to let the user chose the fineness of the tabulation, but since most of the tabulation access is hard-coded, it is a bit of work to implement it cleanly. A less ambitious implementation of only the 124×676 tabulation besides the 46×328 could be easy.
> 3. Interpolation in the table with a fifth order polynomial instead of third order.
>    -> Easy to implement in Capytaine.
> 4. 4-points Gauss quadrature for the integration of the Green function on a face, instead of 1-point approximation.
>    -> Relatively easy to implement in Capytaine. Could be fun to compare all the quadrature implemented in [Quadpy](https://github.com/nschloe/quadpy)
> 5. Irregular frequencies removal.
>    Since it requires the user to mesh the intersection of the body with the free surface (see [Babarit and Delhommeau 2015]), I'm not sure if it was actually active for the results presented previously in this thread.
>    Unlike the previous points, it is a qualitative change in what is computed, not just a more accurate version of the same. Not sure how hard it would be to implement.
> 
> Implementing points 2 to 4 is a nice work program, but I don't know when I'll have the time to do it. However, it is easy enough to be done by new contributors (with some guidance from me). So if anybody wants to volunteer, please let me know.
> 
> Regarding point 5: @izabala123, can I let you check if your Nemoh 115 gives different results when the intersection of the free surface with the body is meshed (see e.g. half-sphere case in [Babarit and Delhommeau 2015])?
> 
> Note that none of this explains the discrepancy between Nemoh 113 and Capytaine, which is what worries me the most.

**Comment by mancellin** [comment]
Hi @ffrancesco 

Thank you for your interest!

Since the last message above, I've added a partial implementation of point 4 in version 1.2 of the code. There is an effect on the results but it seems quite small.

Actually, what is missing the most right now is a set of benchmarks to evaluate the precision gain from changes of the code. Something more quantitative than my eyeball comparisons, for instance some reference problems with analytical solutions. (@izabala123 's test case is interesting but costly to compute and so complex that a lot of things could go wrong.)

If you don't mind theory and literature review, you can help with that. If you are more interested in coding, I'd be happy to point you to the sections of codes that are relevant to point 2.

**Comment by ffrancesco** [comment]
Hi @mancellin I'll have to run some code comparison in the next 3-4 months so I'll be glad to help out with some simpler cases, maybe it will be nice to make a list of cases to run. Further, in the next 12 months I 'll also be happy to contribute to the code development if possible. 
regarding the FWT case, it is possible to try to simplify the structure removing the braces and the heaving plate and make the comparison with other implementation on Nemoh, I'll be happy to help out if @izabala123 can share some details. Regarding the heaving plates: are the dipole implemented in the solver, and if not can that be the problem?

**Comment by mancellin** [comment]
> Hi @mancellin I'll have to run some code comparison in the next 3-4 months so I'll be glad to help out with some simpler cases, maybe it will be nice to make a list of cases to run. Further, in the next 12 months I 'll also be happy to contribute to the code development if possible.

Great!
[This paper](http://mural.maynoothuniversity.ie/12466/) might be a good start to look for interesting test cases. But their might also be plenty of examples with analytical expressions in the literature on linear potential flow.

> regarding the FWT case, it is possible to try to simplify the structure removing the braces and the heaving plate and make the comparison with other implementation on Nemoh, I'll be happy to help out if @izabala123 can share some details. Regarding the heaving plates: are the dipole implemented in the solver, and if not can that be the problem?

Are you refering to @izabala123 's test case? I don't remember it having a thin plate. If so, yes it would be an issue as dipoles are not implemented.

**Comment by ffrancesco** [comment]
Hi @mancellin, thanks for the reference, I'll use that one to start. I'll post an update here when I've any news.

Regarding the semisub platform from @izabala123, you are right that there are no thin plates, but the lower part of the radial columns resemble it and from the image above I do not know if the mesh is refined enough. Regardless, I'd like to play with a similar structure and see if there is a particular element which cause the error.

**Comment by izabala123** [comment]
Dear @ffrancesco 

DeepCWind semisub platform mesh is available [here](https://github.com/izabala123/Nemoh/tree/master/Samples/Nemoh115%20DCW_7000/mesh). You can play with it. Unfortunately I cannot modify the mesh as I do not have now any strong mesher available for personal use.

**Comment by ryancoe** [comment]
For simple test cases, if you can do submerged bodies you can use some analytical solutions to directly measure error in the BEM solution. For example, [Newman](https://mitpress.mit.edu/books/marine-hydrodynamics) gives the solution for a set of spheroids.

**Comment by ffrancesco** [comment]
@ryancoe thanks for the suggestion. I do not have the book on my hands, so i need to find it first. But this for sure a good approach.

**Comment by ffrancesco** [comment]
hi @mancellin, I just noticed that using conda installs capytaine version 1.0.1 and not 1.2, is that correct?

**Comment by mancellin** [comment]
@ffrancesco No, conda should be able to install the latest version. Could you open a separate issue with more details if you can't install 1.2?

**Comment by mancellin** [comment]
> > This is added mass / damping of unit radius hemisphere in heave motion..compared Capytaine with analytical results from "The wave forces acting on a floating hemisphere undergoing forced periodic oscillations By A.HULME". Why is capytaine underrestimating one and overestimating the other?

Thank you for your message. I'm currently trying to improve the accuracy of Capytaine so your input is very welcome. But this message thread is becoming a bit too long. Since your question is not directly related, would you mind opening a new issue? Besides, could you provide the code you used, to help me investigate the issue? In the mean time, I'll take a look at the paper you cited.

**Comment by mancellin** [comment]
Follow-up in #159

**Comment by mancellin** [comment]
The pull requests #156 and #157 seem to have solved the original issues of this thread, that is (a) the mismatch between Capytaine and Nemoh 113 and (b) the discontinuities in the outputs of Capytaine and Nemoh 113 in comparison with the smoother outputs of Wamit and Nemoh 115.

There might still be some work to be done to merge other features of Nemoh 115 in Capytaine (namely: the denser tabulation, the higher order interpolation in the tabulation and the irregular frequencies removal), but it might be easier to discuss them independently than to make this thread even longer.

I think we can now close this issue 🎉
